analysis/token/synonym/synonym.go

```diff
@@ -0,0 +1,378 @@
+package synonym
+
+import (
+	"log"
+
+	"github.com/blevesearch/bleve/v2/analysis"
+	"github.com/blevesearch/bleve/v2/registry"
+	"github.com/blevesearch/bleve/v2/search"
+	"github.com/blevesearch/vellum"
+)
+
+const Name = "synonym"
+
+const SeparatingCharacter = ' '
+
+var equivalentSynonymType = []byte("equivalent")
+var explicitSynonymType = []byte("explicit")
+
+type SynonymFilter struct {
+	Fst              []byte
+	ByteSliceHashMap map[uint64][]byte
+	VellumMap        map[uint64][]uint64
+	Fuzziness        int
+	Prefix           int
+}
+
+type synonymGraphNode struct {
+	token    []byte
+	node     int
+	orig     bool
+	position int
+	start    int
+	end      int
+}
+type fuzzyStruct struct {
+	output uint64
+	state  int
+}
+
+func fuzzySearchFST(output uint64, curr int, depth int, fuzzyQueue []fuzzyStruct, fst *vellum.FST, curWord string, matchTry string, fuzziness int) ([]fuzzyStruct, bool, error) {
+	var fuzzyMatched = false
+	newCur := fst.Accept(curr, SeparatingCharacter)
+	numEdges, err := fst.GetNumTransitionsForState(curr)
+	if err != nil {
+		return nil, false, err
+	}
+	if newCur != 1 || numEdges == 0 {
+		_, exceeded := search.LevenshteinDistanceMax(curWord, matchTry, fuzziness)
+		if !exceeded {
+			fuzzyQueue = append(fuzzyQueue, fuzzyStruct{
+				output: output,
+				state:  curr,
+			})
+			fuzzyMatched = true
+		}
+	}
+	if depth != 0 {
+		edges, err := fst.GetTransitionsForState(curr)
+		if err != nil {
+			return nil, false, err
+		}
+		for _, i := range edges {
+			if i != SeparatingCharacter {
+				newCur, newOut := fst.AcceptWithVal(curr, i)
```

Comment:
no because 'i' is part of edges slice - which is the list of bytes for which transition from curr is defined. So a transition should always be defined for 'i' at curr